File: app/src/main/java/ru/leushinilya/loftmoney/screens/main/budget/BudgetViewModel.java
package ru.leushinilya.loftmoney.screens.main.budget;

import android.content.SharedPreferences;
import android.widget.Toast;

import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;

import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.disposables.Disposable;
import io.reactivex.schedulers.Schedulers;
import ru.leushinilya.loftmoney.LoftApp;
import ru.leushinilya.loftmoney.R;
import ru.leushinilya.loftmoney.cells.Item;
import ru.leushinilya.loftmoney.remote.ItemsAPI;
import ru.leushinilya.loftmoney.remote.RemoteItem;

public class BudgetViewModel extends ViewModel {

    CompositeDisposable compositeDisposable = new CompositeDisposable();
    MutableLiveData<ArrayList<Item>> liveDataItems = new MutableLiveData<>();
    MutableLiveData<String> messageString = new MutableLiveData<>();
    MutableLiveData<Integer> messageInt = new MutableLiveData<>();
    MutableLiveData<Boolean> isEditMode = new MutableLiveData<>(false);

    @Override
    protected void onCleared() {
        super.onCleared();
        compositeDisposable.dispose();
    }

    public void updateListFromInternet(ItemsAPI itemsAPI, int currentPosition, SharedPreferences sharedPreferences) {
        String authToken = sharedPreferences.getString(LoftApp.AUTH_KEY, "");
        String type = "income";
        if (currentPosition == 0) type = "expense";
        Disposable disposable = (itemsAPI.getItems(type, authToken)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(remoteItems -> {
//                        sorting remoteItems list by date
                        Comparator<RemoteItem> comparator = (o1, o2) -> o1.getDate().compareTo(o2.getDate());
                        Collections.sort(remoteItems, comparator);

                        ArrayList<Item> itemList = new ArrayList<>();
                        for (RemoteItem remoteItem : remoteItems) {
                            itemList.add(Item.getInstance(remoteItem));
                        }
                        liveDataItems.postValue(itemList);

                }, throwable -> {
                    messageString.postValue(throwable.getLocalizedMessage());
                }));

        compositeDisposable.add(disposable);
    }

    public void removeItem(Item item, ItemsAPI itemsAPI, SharedPreferences sharedPreferences){
        String authToken = sharedPreferences.getString(LoftApp.AUTH_KEY, "");
        Disposable disposable = itemsAPI
                .removeItem(item.getId(), authToken)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(() -> {
//                    finish();
                }, throwable -> {
                    messageString.postValue(throwable.getLocalizedMessage());
                });
        compositeDisposable.add(disposable);
    }

}

File: app/src/main/java/ru/leushinilya/loftmoney/screens/main/budget/BudgetViewModel.kt
package ru.leushinilya.loftmoney.screens.main.budget

import androidx.lifecycle.ViewModel
import io.reactivex.disposables.CompositeDisposable
import androidx.lifecycle.MutableLiveData
import ru.leushinilya.loftmoney.remote.ItemsAPI
import android.content.SharedPreferences
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateListOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import ru.leushinilya.loftmoney.LoftApp
import io.reactivex.disposables.Disposable
import io.reactivex.schedulers.Schedulers
import io.reactivex.android.schedulers.AndroidSchedulers
import ru.leushinilya.loftmoney.cells.Item
import ru.leushinilya.loftmoney.remote.RemoteItem
import java.util.*

class BudgetViewModel : ViewModel() {
    var compositeDisposable = CompositeDisposable()
    var items by mutableStateOf(listOf<Item>())
    var messageString = MutableLiveData<String>()
    var messageInt = MutableLiveData<Int>()
    var isEditMode = MutableLiveData(false)
    @JvmField
    var incomesSum = MutableLiveData(0f)
    @JvmField
    var expensesSum = MutableLiveData(0f)
    override fun onCleared() {
        super.onCleared()
        compositeDisposable.dispose()
    }

    fun updateListFromInternet(
        itemsAPI: ItemsAPI,
        currentPosition: Int,
        sharedPreferences: SharedPreferences
    ) {
        val authToken = sharedPreferences.getString(LoftApp.AUTH_KEY, "")
        var type = "income"
        if (currentPosition == 0) type = "expense"
        val disposable = itemsAPI.getItems(type, authToken)
            .subscribeOn(Schedulers.io())
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe({ remoteItems: ArrayList<RemoteItem> ->
//                        sorting remoteItems list by date
                val comparator =
                    Comparator { o1: RemoteItem, o2: RemoteItem -> o1.date.compareTo(o2.date) }
                Collections.sort(remoteItems, comparator)
                val itemList = ArrayList<Item>()
                var sum = 0f
                for (remoteItem in remoteItems) {
                    itemList.add(Item.getInstance(remoteItem))
                    sum += remoteItem.price
                }
                items = itemList
                if (currentPosition == 0) expensesSum.postValue(sum) else if (currentPosition == 1) incomesSum.postValue(
                    sum
                )
            }) { throwable: Throwable -> messageString.postValue(throwable.localizedMessage) }
        compositeDisposable.add(disposable)
    }

    fun removeItem(item: Item, itemsAPI: ItemsAPI, sharedPreferences: SharedPreferences) {
        val authToken = sharedPreferences.getString(LoftApp.AUTH_KEY, "")
        val disposable = itemsAPI
            .removeItem(item.id, authToken)
            .subscribeOn(Schedulers.io())
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe({}) { throwable: Throwable -> messageString.postValue(throwable.localizedMessage) }
        compositeDisposable.add(disposable)
    }
}